backend/src/test/java/com/google/minicrm/servlets/FormsServletTest.java

```diff
@@ -0,0 +1,283 @@
+    verify(response, never()).sendError(Mockito.anyInt());
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_throws400WithNoFields() throws Exception {
+    when(request.getContentType()).thenReturn("application/x-www-form-urlencoded;");
+    formsServlet.doPut(request, response);
+
+    //verify the response has an error
+    verify(response).sendError(eq(400), Mockito.anyString()); //we want an error msg
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_throws400WithNoBody() throws Exception {
+    when(request.getContentType()).thenReturn("application/json;");
+    BufferedReader br = new BufferedReader(new StringReader(""));
+    when(request.getReader()).thenReturn(br);
+    formsServlet.doPut(request, response);
+
+    //verify the response has an error
+    verify(response).sendError(eq(400), Mockito.anyString()); //we want an error msg
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_throws400BlankParameter() throws Exception {
+    when(request.getContentType()).thenReturn("application/x-www-form-urlencoded;");
+    when(request.getParameter("formId")).thenReturn("");
+    when(request.getParameter("formName")).thenReturn("name");
+    formsServlet.doPut(request, response);
+
+    //verify the response has an error
+    verify(response).sendError(eq(400), Mockito.anyString()); //we want an error msg
+  }
+
+  @Test
+  public void formsServlet_putRequest_json_throws400BlankParameter() throws Exception {
+    when(request.getContentType()).thenReturn("application/json;");
+    Reader reader = new StringReader(new FormsPutRequest("", "newName").toJson());
+    when(request.getReader()).thenReturn(new BufferedReader(reader));
+    formsServlet.doPut(request, response);
+
+    //verify the response has an error
+    verify(response).sendError(eq(400), Mockito.anyString()); //we want an error msg
+  }
+
+  @Test
+  public void formsServlet_putRequest_throws415OnInvalidContentType() throws Exception {
+    when(request.getContentType()).thenReturn("multipart/form-data;");
+    formsServlet.doPut(request, response);
+
+    //verify the error with a message
+    verify(response).sendError(eq(415), Mockito.anyString());
+  }
+
+  /**
+   * Seeds forms but with one second delays inbetween because datastore only has a 1 second
+   * resolution for date objects
+   * @throws Exception
+   */
+  private void slowSeedForms() throws Exception {
+    User testUser = new User("email", "authDomain", TEST_USER_ID);
+    Key parentKey = Advertiser.generateKey(testUser);
+    form1 = new Form(parentKey, 1, "form1");
+    TimeUnit.SECONDS.sleep(1); //datastore timestamp only has 1 second precision
+    form2 = new Form(parentKey, 2, "form2");
+    TimeUnit.SECONDS.sleep(1);
+    form3 = new Form(parentKey, 3, "form3");
+    TimeUnit.SECONDS.sleep(1);
+    DatastoreUtil.put(form1);
+    DatastoreUtil.put(form2);
+    DatastoreUtil.put(form3);
+  }
```

Comment:
Needed to do this since the datastore only has a 1 second precision meaning that I wouldn't be able to test the forms being returned in the correct order without this. Is this implementation fine (its a bit slow since it forces a 3s runtime on one of the tests) or should I change the Forms object and allow myself to set the date instance variable itself?